### PR TITLE
Initial commit -lots of small changes to make the sln compilable 

### PR DESCRIPTION
**Pull request by patrickac**
Complete Exercise 0101 with unit tests
Make corrections for compilability

### CODE REVIEW

File: InterviewPrep/chapter01/question0101.cs [hunk 1]
```diff
@@ -1,26 +1,91 @@
...
+                for (var j = i + 1; j < s.Length; j++)
+                {
+                    if (s[i] == s[j])
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
```
Comment (rsalota) [inline InterviewPrep/chapter01/question0101.cs]: 1. Assuming that the character set is fixed - can we do something better than N^2 solution . I am looking for an o(N) solution.

File: InterviewPrep/chapter01/question0101.cs [hunk 2]
```diff
@@ -1,26 +1,91 @@
...
+            {
+                return true;
+            }
+
+            var cd = new ConcurrentDictionary<char, bool>();
+            var unique = true;
+            Parallel.ForEach(s, c =>
+                    {
+                        if (cd.ContainsKey(c))
+                        {
```
Comment (rsalota) [inline InterviewPrep/chapter01/question0101.cs:38]: One comment is on using multithreading here - multi threading is not really useful for CPU intensive problems. The cost of spinning a thread in this case might be too expensive. 

In general the code looks good.

File: InterviewPrep/chapter01/question0101.cs [hunk 3]
```diff
@@ -1,26 +1,91 @@
...
+        /// <summary> Does the string have unique characters.    </summary>
+        /// <param name="s"> The s.   </param>
+        /// <exception cref="AggregateException"> The exception that contains all the individual exceptions thrown on all threads.  </exception>
+        /// <exception cref="OverflowException"> The dictionary already contains the maximum number of elements,     <see cref="F:System.Int32.MaxValue"/>. </exception> 
+        /// <exception cref="ArgumentNullException"> <paramref name="key"/> is a null reference (Nothing in Visual Basic).  </exception>
+        /// <returns> The <see cref="bool"/>.  </returns>
+        public static bool HasUniqueCharacters(string s)
+        {
+            // Contract.Requires(s != null);
+            if (CornerCase(s))
```
Comment (rsalota) [inline InterviewPrep/chapter01/question0101.cs:28]: This is a good pattern .. keep doing it .

File: InterviewPrep/chapter01/question0103.cs [hunk 4]
```diff
@@ -1,19 +1,98 @@
...
+        /// <exception cref="ArgumentException">
+        /// </exception>
+        /// <exception cref="ArgumentNullException">
+        /// <paramref name="array"/> is null.
+        /// </exception>
+        /// <exception cref="InvalidOperationException">
+        /// One or more elements in <paramref name="array"/> do not implement the
+        ///     <see cref="T:System.IComparable`1"/> generic interface.
+        /// </exception>
+        public static bool IsPermutation(string s1, string s2)
```
Comment (rsalota) [inline InterviewPrep/chapter01/question0103.cs:43]: What is the complexity of this solution? The idea is to try to do this in the most efficient way ?

File: InterviewPrep/chapter01/question0103.cs [hunk 5]
```diff
@@ -1,19 +1,98 @@
...
         {
+            if (CornerCase(s1, s2))
```
Comment (rsalota) [inline InterviewPrep/chapter01/question0103.cs:45]: Not sure why when s.length==1 will this lead to a true without inspecting equality? One thing i suggest is to try to dry run a few test cases to check these edge cases.

**Comment by rsalota** [comment]
Thanx
